core/kotlinx-coroutines-test/src/TestCoroutineDispatcher.kt

```diff
@@ -0,0 +1,296 @@
+
+    /**
+     * Moves the current virtual clock forward just far enough so the next delay will return.
+     *
+     * @return the amount of delay-time that this Dispatcher's clock has been forwarded.
+     */
+    @ExperimentalCoroutinesApi
+    fun advanceTimeToNextDelayed(): Long
+
+    /**
+     * Immediately execute all pending tasks and advance the virtual clock-time to the last delay.
+     *
+     * @return the amount of delay-time that this Dispatcher's clock has been forwarded.
+     */
+    @ExperimentalCoroutinesApi
+    fun advanceUntilIdle(): Long
+
+    /**
+     * Run any tasks that are pending at or before the current virtual clock-time.
+     *
+     * Calling this function will never advance the clock.
+     */
+    @ExperimentalCoroutinesApi
+    fun runCurrent()
+
+    /**
+     * Call after a test case completes.
+     *
+     * @throws UncompletedCoroutinesError if any pending tasks are active, however it will not throw for suspended
+     * coroutines.
+     */
+    @ExperimentalCoroutinesApi
+    @Throws(UncompletedCoroutinesError::class)
+    fun cleanupTestCoroutines()
+
+    /**
+     * Run a block of code in a paused dispatcher.
+     *
+     * By pausing the dispatcher any new coroutines will not execute immediately. After block executes, the dispatcher
+     * will resume auto-advancing.
+     *
+     * This is useful when testing functions that that start a coroutine. By pausing the dispatcher assertions or
+     * setup may be done between the time the coroutine is created and started.
+     */
+    @ExperimentalCoroutinesApi
+    suspend fun pauseDispatcher(block: suspend () -> Unit)
+
+    /**
+     * Pause the dispatcher.
+     *
+     * When paused the dispatcher will not execute any coroutines automatically, and you must call [runCurrent], or one
+     * of [advanceTimeBy], [advanceTimeToNextDelayed], or [advanceUntilIdle] to execute coroutines.
+     */
+    @ExperimentalCoroutinesApi
+    fun pauseDispatcher()
+
+    /**
+     * Resume the dispatcher from a paused state.
+     *
+     * Resumed dispatchers will automatically progress through all coroutines scheduled at the current time. To advance
+     * time and execute coroutines scheduled in the future use one of [advanceTimeBy], [advanceTimeToNextDelayed],
+     * or [advanceUntilIdle].
+     */
+    @ExperimentalCoroutinesApi
+    fun resumeDispatcher()
+}
+
+/**
+ * Thrown when a test has completed by there are tasks that are not completed or cancelled.
+ */
+@ExperimentalCoroutinesApi
+class UncompletedCoroutinesError(message: String, cause: Throwable? = null): AssertionError(message, cause)
+
+/**
+ * [CoroutineDispatcher] that can be used in tests for both immediate and lazy execution of coroutines.
+ *
+ * By default, [TestCoroutineDispatcher] will be immediate. That means any tasks scheduled to be run without delay will
+ * be immediately executed. If they were scheduled with a delay, the virtual clock-time must be advanced via one of the
```

Comment:
this behavior is a bit worrying.
Specifically, an asyc being dispatched immediately significantly alters my coroutine's execution order. Of course, there are no guaratees but  consider a buggy code like this:

```
val atomic = AtomicBoolean(false)
async {
    if (fetchData() == true) {
        atomic.set(true)
    }
}
// do something else
if (atomic.set()) {
    // do something.
}
``` 
normally, the code above would flake since `async` may or may not be scheduled immediately. By altering that behavior and making it consistent; this might hide bugs in applications. If we need to pick a consistent behavior, i would prefer asyncs and launches to be deferred as much as possible (put them end of the execution queue). 
That is not perfect either but i think it is better behavior since it will make it easier to catch any missing `await` calls for side effects.

Reply:
This is one I've gone back and forth on - in this specific example it would be prudent for a developer to write the test like this

```
- test of delayed async call
runBlockingTest {
    pauseDispatcher {
         subject()
         runCurrent()
    }
}

- test of early async call
runBlockingTest {
    subject()
}
```

I modeled the current behavior off the Robolectric scheduler which typically works for most code that doesn't introduce either race conditions or inappropriate side-effects at launch. For code that does, `pauseDispatcher` provides the ability to engage in explicit ordering control.

Eager execution in tests provides a benefit over the execution for tests for this typical code:

```
fun subject() { // don't return job
    scope.launch { // stuff }
}

@Test
fun testSubject() = scope.runBlockingTest {
    subject()
}
```

In execution in the `runBlocking` style, `launch` body won't be entered until the `job.join()`, which isn't available. With this dispatcher, the test could call `runCurrent()` after `subject()` which is similar to the Rx implementation.

As another detail: All of this only applies to launched coroutines. Regular suspend functions will automatically progress through both dispatch and delay (calling another suspend function or a delay function) in `runBlockingTest` (since they don't start a coroutine there is no way to return control back to the test case without inventing new semantics).

Again, I have gone back and forth on this a few times and would welcome other ideas.  One way to go would be to invert the behavior, and expose a way to enter auto-progressing dispatcher state (maybe `resumeDispatcher` could take a block?)